broker/lib/bosh/BoshOperationCache.js

```diff
@@ -0,0 +1,243 @@
+   */
+  getBoshTask(serviceInstanceId) {
+    logger.debug(`Getting bosh task ID for service instance id ${serviceInstanceId}`);
+    const wrapper = new Promise((resolve, reject) => {
+      const key = getTaskKey(serviceInstanceId);
+      etcdConnector().get(key).string().then(obj => {
+        resolve(obj);
+      }).catch(err => reject(err));
+    });
+    return wrapper;
+  }
+
+  /**
+   * Fetches parameters for a BOSH deployment operation from the Etcd store based on the name
+   * 
+   * @param {Deployment Name} name 
+   */
+  getDeploymentByName(name) {
+    logger.debug('Getting deployment for ', name);
+    const wrapper = new Promise((resolve, reject) => {
+      etcdConnector().get(getKey(name)).json().then(out => resolve(out)).catch(err => reject(err));
+    });
+    return wrapper;
+  }
+
+  /**
+   * Deletes a single BOSH deployment from the Etcd Store
+   * 
+   * @param {Deployment Name} deploymentName 
+   */
+  deleteDeploymentFromCache(deploymentName) {
+    logger.info('Removing deployment from cache', deploymentName);
+    const wrapper = new Promise((resolve, reject) => {
+      etcdConnector().delete().key(getKey(deploymentName)).then(out => resolve(out)).catch(err => reject(err));
+    });
+    return wrapper;
+  }
+
+  /**
+   * Deletes multiple BOSH deployment names from the Etcd store
+   * 
+   * @param {Deployment Names} deployments 
+   */
+  deleteDeploymentsFromCache(...deployments) {
+    const deploymentIdentifiers = _.flattenDeep(deployments);
+    logger.info('Removing multiple deployments from cache', deployments);
+    const deletePromises = _.map(deploymentIdentifiers, this.deleteDeploymentFromCache);
+    return Promise.all(deletePromises);
+  }
+
+  /**
+   * Stores the BOSH task ID corresponding to the service instance ID when the task is submitted to the BOSH Director
+   * 
+   * @param {Service Instance ID} serviceInstanceId 
+   * @param {BOSH Task ID corresponding to the operation} taskId 
+   */
+  storeBoshTask(serviceInstanceId, taskId) {
+    logger.info("Putting the current deployment task into the cache", serviceInstanceId, taskId);
+    const wrapper = new Promise((resolve, reject) => {
+      const key = getTaskKey(serviceInstanceId);
+      etcdConnector().put(key).value(taskId)
+        .then(() => resolve(true))
+        .catch(err => {
+          logger.error('Error in storing BOSH task id for instance', err);
+          reject(err);
+        });
+    });
+    return wrapper;
+  }
+
+  /**
+   * Stores the deployment operation (user-triggered) into the Etcd store as a JSON document
+   * Key is prefixed with a specific format and the deployment name is used as the identifier
+   * 
+   * @param {Service plan ID for the service instance} planId 
+   * @param {Calculated bosh deployment name} deploymentName 
+   * @param {Parameters used for the operation} params 
+   * @param {Arguments used for the operation} args 
+   */
+  store(planId, deploymentName, params, args) {
```

Comment:
just trying to understand , since all BOSH based operations will put the instance/deployment in "In progress" state user cannot trigger lets say create followed by update or delete etc.. I think hence the current key structure should work.

Similarly thing should hold true for fabrik initiated operations.

maybe we need a differentiator between user triggered vs fabrik triggered for the same deployment . But even here since there is a lock on the instance, dont see both occurring in tandem.

Reply:
But there could be cases where we will not have mutually exclusive operations for example bind and backup are both read operations as per the SF 2.0 locking implementation. If these are the 2 requests which would simultaneously arrive would not have same key name be a problem.

Reply:
But, bind and backup never go via the Director. Exclusively, operations that create bosh tasks will be limited via the concept. Everything else happens as is